### ISSUE TITLE
about leap seconds

### ISSUE THREAD

**Issue by BurntSushi** [BurntSushi/jiff#7]
### What should you do if you want leap second support in Jiff?

The most effective thing you can do is leave a comment in this issue with a _detailed_ account of why you want leap second support. Leaving a comment like, "I need to-the-second precision when calculating differences between instants" is not sufficient. You should say _why_ you need that level of precision, and especially, what would happen if you didn't have it. You should also share what your current solution is and any relevant "industry standards" in your domain. (So for example, if your use case is, "I need to do astronomical math," then it would be nice to say why you can't or don't want to use a specialized library for this purpose.) The most important thing you can do is **connect specific support of leap seconds to an end user experience.** Nebulous "I want it because it's correct" comments are not especially helpful.

The rest of this issue does a deep dive on what leap second support means and various API designs.

-----

**NOTE:** When I originally wrote this, `Timestamp` was called `Instant`. I tried replacing references to `Instant` with `Timestamp` below, but a timestamp is still referred to as an "instant" in places. This concept is still correct. A `Timestamp` _is_ an instant. The only thing that changed was the type name in Jiff.

-----

[This required background reading](http://www.madore.org/~david/computers/unix-leap-seconds.html) gives a nice overview of the relationship between different time scales. (This also makes a proposal at the end, but that's not particularly relevant here. The content leading up to the proposal though I found quite helpful in understanding things.)

Here is a bit [more background reading](https://unix.stackexchange.com/questions/283164/unix-seconds-tai-si-seconds-leap-seconds-and-real-world-code), which serves as a nice way of observing that pretty much everybody gets this wrong in some way. And the reason why we get it wrong pretty much comes down to the fact that we like to represent time as "seconds since an epoch."

And finally, this [reference about Time Standards](https://geometrian.com/programming/reference/timestds/index.php) I found to be a bit ranty, but also very informative. It's nice because it starts from first principles and builds up everything from there.

# Initial thoughts and ecosystem comparison

When I initially started developing Jiff, I had a goal to provide support for leap seconds. The failure to support leap seconds is a common indictment of datetime libraries, and I wanted to avoid falling into that trap. Roughly speaking, "support" in my mind looked like this. And note that when I say `Timestamp`, it also applies to `Zoned` (but not necessarily `civil::DateTime`).

1. Parsing from a string into a `Timestamp` like `2016-12-31T23:59:60Z` (the latest leap second to have occurred as of 2024-04-06) would correctly represent this time as distinct from all other times, including `2016-12-31T23:59:59Z`.
2. Converting an `Timestamp` back into a string would correctly preserve the leap second.
3. Computing the difference between two `Timestamp`s would correctly account for leap seconds. For example, `2017-01-01T00:00:00Z - 2016-12-31T23:00:00` would yield `3601` seconds instead of `3600` seconds.

Existing Rust datetime libraries have pretty spotty support for all of the above. Let's test each of the above properties in isolation for each of `chrono == 0.4.34`, `time == 0.3.34` and `hifitime == 3.9.0`. First, can the library handle parsing a leap second such that the datetime values are distinct?

```rust
fn leap_second_parsing() {
    let leap = "2016-12-31T23:59:60Z"
        .parse::<chrono::DateTime<chrono::Utc>>()
        .unwrap();
    let nonleap = "2016-12-31T23:59:59Z"
        .parse::<chrono::DateTime<chrono::Utc>>()
        .unwrap();
    // `chrono` passes this test.
    assert_ne!(leap, nonleap);

    // For `time`, leap seconds are rejected when parsing
    // via the ISO 8601 format, but are supported via the
    // RFC 2822 format. We want to be maximally charitable
    // here, so pick a format that supports leap seconds.
    let leap = time::OffsetDateTime::parse(
        "Sat, 31 Dec 2016 23:59:60 GMT",
        &time::format_description::well_known::Rfc2822,
    )
    .unwrap();
    let nonleap = time::OffsetDateTime::parse(
        "Sat, 31 Dec 2016 23:59:59 GMT",
        &time::format_description::well_known::Rfc2822,
    )
    .unwrap();
    // `time` passes this test.
    assert_ne!(leap, nonleap);

    let leap = "2016-12-31T23:59:60 UTC".parse::<hifitime::Epoch>().unwrap();
    let nonleap =
        "2016-12-31T23:59:59 UTC".parse::<hifitime::Epoch>().unwrap();
    // `hifitime` fails this test. Interestingly, both of the above times parse
    // as `2016-12-31T23:59:57 UTC`, which seems like a bug. Possibly related:
    // https://github.com/nyx-space/hifitime/issues/255
    // https://github.com/nyx-space/hifitime/issues/288
    assert_eq!(leap, nonleap);
}
```

Only `chrono` and `time` gets this particular case right.

Next is what happens when we try to format an instant corresponding to a leap second. This one is tricky to test in isolation because it requires the library itself to be at least minimally aware of leap seconds as a concept. If, for example, the library _only_ deals with Unix time (which is what `std::time::SystemTime` is), then it's impossible for the library to get this case correct. That's because Unix time is ambiguous with respect to leap seconds, in exactly the same way the civil datetimes are ambiguous with respect to DST transitions. For example, the Unix timestamp `1483228799` could correspond to `2016-12-31T23:59:59Z` or to `2016-12-31T23:59:60Z`. There is quite literally no way to distinguish between them. Analogously, `2024-11-03T1:30:00[America/New_York]` could correspond to either `2024-11-03T05:30:00Z` or to `2024-11-03T06:30:00Z` because of the jump backwards from DST to standard time. (This analogy between leap second and DST handling will prove to be prescient.)

In other words, to get the second property right, you kind of need to pass the first property. That is, both properties are required in order to correctly roundtrip datetimes. Since `hifitime` didn't satisfy the first property, we will manually create an instant corresponding to a leap second (which `hifitime` explicitly supports by virtue of representing its instants in TAI) to test whether it can format the leap second correctly:

```rust
fn leap_second_formatting() {
    let leap = "2016-12-31T23:59:60Z"
        .parse::<chrono::DateTime<chrono::Utc>>()
        .unwrap();
    let formatted = leap.to_string();
    // `chrono` passes this test.
    assert_eq!(formatted, "2016-12-31 23:59:60 UTC");

    let leap = time::OffsetDateTime::parse(
        "Sat, 31 Dec 2016 23:59:60 GMT",
        &time::format_description::well_known::Rfc2822,
    )
    .unwrap();
    let formatted =
        leap.format(&time::format_description::well_known::Rfc2822).unwrap();
    // `time` fails this test because it clamps 23:59:60
    // to 23:59:59.999999999 upon parsing. And there is no
    // other way (AFAIK) to represent a leap second in a `time`
    // data type.
    assert_ne!(formatted, "Sat, 31 Dec 2016 23:59:60 +0000");

    // `hifitime` seems to fail this test as well. The string it prints is
    // `2016-12-31T23:59:58 UTC`.
    let leap =
        hifitime::Epoch::from_gregorian_utc(2016, 12, 31, 23, 59, 60, 0);
    assert_ne!(leap.to_string(), "2016-12-31T23:59:60 UTC");
}
```

`chrono` passes this test, but neither `time` nor `hifitime` do. `hifitime`'s documentation seems to suggest that this it should satisfy this property (and the first), but neither seem to be the case here.

The third property is a bit different from the first two. The first two deal with the case where the leap second hits you directly. That is, the time being dealt with is itself a leap second. This is somewhat of a particularly rare case because POSIX/Unix time is pretty ubiquitous and POSIX/Unix time simply ignores leap seconds altogether. So it's pretty rare to see something like `2016-12-31T23:59:60Z` in the wild, even when it accurately and precisely describes the current instant in time. Instead, you'll just see `2016-12-31T23:59:59Z`.

The third property however tests the handling of leap seconds _implicitly_. That is, even when two times aren't themselves leap seconds, does computing the difference between them account for leap seconds if one or more of them occurred in the interval of time between them? Let's see how the various Rust libraries do on this test:

```rust
fn leap_second_difference() {
    let t1 = chrono::NaiveDateTime::new(
        chrono::NaiveDate::from_ymd_opt(2017, 1, 1).unwrap(),
        chrono::NaiveTime::from_hms_opt(0, 0, 0).unwrap(),
    );
    let t2 = chrono::NaiveDateTime::new(
        chrono::NaiveDate::from_ymd_opt(2016, 12, 31).unwrap(),
        chrono::NaiveTime::from_hms_opt(23, 0, 0).unwrap(),
    );
    // `chrono` fails, returns 3600 seconds.
    assert_ne!(t1 - t2, chrono::TimeDelta::seconds(3601));

    let t1 = time::PrimitiveDateTime::new(
        time::Date::from_calendar_date(2017, time::Month::January, 1).unwrap(),
        time::Time::from_hms(0, 0, 0).unwrap(),
    );
    let t2 = time::PrimitiveDateTime::new(
        time::Date::from_calendar_date(2016, time::Month::December, 31)
            .unwrap(),
        time::Time::from_hms(23, 0, 0).unwrap(),
    );
    // `time` fails, returns 3600 seconds.
    assert_ne!(t1 - t2, time::Duration::seconds(3601));

    let t1 = hifitime::Epoch::from_gregorian_utc(2017, 1, 1, 0, 0, 0, 0);
    let t2 = hifitime::Epoch::from_gregorian_utc(2016, 12, 31, 23, 0, 0, 0);
    // `hifitime` passes!
    assert_eq!(t1 - t2, hifitime::Duration::from_seconds(3601.0));
}
```

Interesting! Despite not satisfying properties 1+2, `hifitime` does satisfy property 3. And in particular, neither `chrono` nor `time` satisfy property 3. This is likely because `hifitime` _should_ satisfy properties 1+2, but perhaps there are some bugs in its implementation. Conversely, I do not believe the internal design of either `chrono` or `time` allows it to satisfy property 3 easily.

The salient difference between `hifitime` and `chrono`/`time` is that `hifitime`'s representation of time is TAI. That is, [International Atomic Time](https://en.wikipedia.org/wiki/International_Atomic_Time). TAI doesn't have leap seconds. It just always ticks forward without regard to whether it's in sync with the rotation of the Earth. This means that you can subtract any two points in TAI time and get a precise answer as to how many seconds elapsed between them. The way this is done in `hifitime` (at present) is, as I understand it, as follows:

* It represents a table of leap seconds as defined by IERA (and included in the IANA Time Zone Database, e.g., `/usr/share/zoneinfo/leapseconds`).
* When converting from Unix time (i.e., what `std::time::SystemTime` is) or UTC to an `Epoch`, it "applies" the leap seconds relevant for that particular instant in time. So for example, to convert `2016-12-31T23:59:59Z` to TAI, you'd add 37 seconds. (That's 27 leap seconds added since 1972 plus the initial 10 second difference definitionally declared on 1972-01-01 between TAI and UTC.) That is, the internal representation of an `Epoch` is just TAI. (Or more precisely, the amount of time that has passed from its chosen epoch, `1900-01-01T00:00:00`.)
* Whenever you need to get a datetime back from an `Epoch`, you need to subtract the leap seconds from TAI to get UTC back.

This makes it somewhat clearer why properties 1+2 are perhaps not as well tested (or as high priority) in `hifitime`. Namely, they only really apply when dealing with a leap second directly, where as property 3 applies much more frequently. The only thing that needs to be true for property 3 to matter is for at least one leap second to have occurred between two instants.

The downside of `hifitime`'s approach is that it needs to do a [non-trivial calculation](https://github.com/nyx-space/hifitime/blob/822c0a0ee4da21f385bc520547431fd1b3ed7d6f/src/epoch.rs#L245-L263) every single time an `Epoch` is created from a Unix timestamp _and_ every single time an `Epoch` is converted to a civil datetime (which `hifitime` doesn't have).

# Observations

After digesting the above, I went and looked at what other datetime libraries. From what I could tell, there are essentially _zero_ general purpose datetime libraries that deal with leap seconds in a way that satisfies all three of the above properties. This includes Joda Time (now `java.util.time`) and the in-development Temporal library. In particular, Temporal say [this](https://tc39.es/proposal-temporal/docs/instant.html):

> Like Unix time, Temporal.Instant ignores leap seconds.

and [this](https://tc39.es/proposal-temporal/docs/zoneddatetime.html)

> NOTE: Although Temporal does not deal with leap seconds, dates coming from other software may have a second value of 60. In the default 'constrain' mode and when parsing an ISO 8601 string, this will be converted to 59. In 'reject' mode, this function will throw, so if you have to interoperate with times that may contain leap seconds, don't use reject.

on the topic of leap seconds.

Notably, C++'s standard `chrono` library _does_ support leap seconds by introducing a "clock" abstraction where one can convert between system (Unix) time and UTC. That conversion step essentially does the same leap second calculation that `hifitime` does.

Of course, there are other more _specialized_ datetime libraries that do support leap seconds. For example, [Time4J](https://github.com/MenoData/Time4J).

Given that Temporal was my principal influence in the design of this library, I wanted to defer strongly to them. Still, though, `hifitime`'s approach, to me, made it look like supporting leap seconds wasn't that crazy. On the other hand, I didn't want to introduce the overhead of dealing with leap seconds for almost all operations on a Jiff `Timestamp`...

# Making `Timestamp` generic over a `TimeScale`

So my initial plan to deal with leap seconds was to make the `Timestamp` type generic over a trait called `TimeScale`. I perceived this as pretty similar to what `chrono` does in C++. The default would be the `Unix` time scale. That is, leap seconds are completely ignored. This may not produce exactly correct result when computing the difference between two instants, but it is synergistic with how most system clocks work. And there wouldn't be any additional overhead.

The idea then was that users of Jiff could opt into leap second handling by using the `TAI` `TimeScale` instead of `Unix`. In so doing, any construction of an `Timestamp` from Unix time would require doing the same sort of leap second calculation that `hifitime` does. And then whenever one needs to convert back to a civil datetime, the same calculation would be done in reverse. In exchange, subtracting two `Timestamp` values in the `TAI` time scale would produce precise results. Unlike `hifitime`, this design wouldn't permit subtracting _any_ two `Timestamp` values regardless of their time scale because the internal representation would be defined by the `TimeScale`. Indeed, this is the critical difference from `hifitime`'s design that permits us to make leap second support opt-in without the overhead of always needing to deal with it. The downside is that you don't get any kind of interoperability between instants of different time scales. But this seemed fine to me, because you could always explicitly convert between time scales.

I was not a huge fan of this design because it adds a type parameter to a foundational type, and has a somewhat infectious effect. That is, anything that deals with `Timestamp`s now needs to also consider the time scale in its API. I found this to be annoying to deal with even internally. For example, I wanted to use `Timestamp<Unix>` to represent timestamps from TZif files (indicating the start of DST transitions in `Unix` time), but the API wants to accept any generic `Timestamp<S>`. Since my design means you can't compare instants of different time scales, every time zone lookup would in turn need to do a leap second conversion. This was bad juju to me and finally convinced me to get rid of this approach.

(There does exist the `/usr/share/zoneinfo/right` database of TZif files which defines its transitions in `TAI-10` time that could potentially be used instead. But dealing with this is a pretty complex mess, because you'd sometimes want to use the "normal" TZif files and sometimes want to use the "right" TZif files, and which one you pick depends on the `TimeScale` of the instant you have. And what happens when the `right` directory doesn't exist?)

# Status quo

At time of writing, I ripped out the `TimeScale` abstraction. An `Timestamp` is now always Unix time. (And I'm now considering renaming it to `Timestamp` because I don't like that it has the same name as `std::time::Instant`.)

I haven't done parsing/formatting yet, but at present, I still currently plan to adopt Temporal's appoach _at minimum_: if a leap second is seen, then clamp it into a valid time.

But that still leaves the question of how to deal with subtracting two times that cross a leap second discontinuity and still get a correct result. It is my belief that this is quite a challenging problem, and not even `hifitime` gets it fully correct. Or I should say, it doesn't provide the right user experience. In particular, one nice thing about Jiff (and Temporal, which inspired this API) is that you can pretty seamlessly work with durations in any unit of time. For example, you can add things like "years" or even get "years" as a unit of difference between two datetimes. However, if our instants accounted for leap seconds, then `2017-03-01T00:00:00Z - 2016-03-01T00:00:00Z` would give you a result of `1 year 1 second`. I frankly find that to be a bit weird, and is analogous to `2024-11-04T00:00:00-05[America/New_York] - 2024-11-03T00:00:00-04[America/New_York]` yielding `1 day 1 hour` instead of the more intuitive and typically desired `1 day`. Indeed, [this is what Temporal does](https://tc39.es/proposal-temporal/docs/zoneddatetime.html#subtract).

It's worth lingering on this point for a bit, because the way Temporal works here is very intuitive. Namely, when adding units _less_ than a day (hours, minutes, seconds, milliseconds, microseconds, nanoseconds) even when they add up to move than a day, then Temporal will incorporate DST transitions. For example, this program:

```javascript
let dt = Temporal.ZonedDateTime.from({
  timeZone: 'America/New_York',
  year: 2024,
  month: 3,
  day: 10,
  hour: 1,
  minute: 30,
});
let duration = Temporal.Duration.from({hours: 24});
console.log(dt.add(duration).toString());
```

Has this output:

```
2024-03-11T02:30:00-04:00[America/New_York]
```

But if we change the duration to `days: 1` instead of `hours: 24`, we get this output:

```
2024-03-11T01:30:00-04:00[America/New_York]
```

In other words, units of "day" or higher are of variable length depending on the datetime they are interpreted relative to. But units below "day" are always of a fixed length.

 I think `hifitime` doesn't really have this "problem" because it doesn't have the same sort of intuitive interface for representation durations of time. That is, its unit of "day" is a fixed length of 86,400 seconds _always_. Indeed, `hifitime` doesn't deal with DST at all. So while `hifitime` gets some part of leap second handling correct (even if we ignore it not satisfying properties 1+2 above as "just bugs"), it doesn't seem like a foundation one can build on to provide a nice user experience like what Temporal does. Indeed, I believe this also applies to C++'s `chrono` library, but my C++ knowledge is pretty weak and thus can't be sure of what C++'s `chrono`'s higher level user experience is like.

One way to look at the issue of leap seconds is that it is strictly analogous to DST handling:

* A positive leap second creates a fold in time, just like a time transition that moves the clocks backwards does. (This is usually, but not always, the transition from DST to standard time.)
* A negative leap second creates a gap in time, just like a time transition that moves the clocks forwards does. (This is usually, but not always, the transition from standard time to DST.)

In order to make this work at the user experience level, Temporal had to make any units above "hours" variable length. This is because all time transitions are expressed in units of hours (with a possible fractional component). So once you move above hours, it makes sense to "ignore" the time transitions. For example, if you have an appointment on `2024-03-09T14:00:00-05[America/New_York]` and want to re-schedule it to one day later, then the expected appointment time would be `2024-03-10T14:00:00-04[America/New_York]`. If you implemented that re-scheduling by simply adding 24 hours, then you'd get the wrong result of `2024-03-10T15:00:00-04[America/New_York]`.

So if we wanted to deal with leap seconds in analogous way, this appears to lead us to the conclusion that any units above _seconds_ need to be of variable length. That means, for example, that adding 1 minute to `2016-12-31T23:59:59Z` should give you `2017-01-01T00:00:59Z`, while adding 60 seconds should give you `2017-01-01T00:00:58Z`. Wild, right?

From what I can tell, the division between which units are variable (days and higher) and which are not (hours and lower) in Temporal is an architectural constraint the pervades not just its API but also its implementation. For example, if you can assume that all units of hours or below are of a fixed length, then arithmetic using hours or below can be performed by simple addition on timestamps. At the API level, Temporal permits rounding durations to the nearest hour (or below) without specifying a `relativeTo` datetime. But if rounding days or above, since days and above are variable sized units, the operation necessarily requires a datetime to interpret the duration in relation to. Otherwise, it can't actually know how long a "day" is.

While these constraints appear easy to mechanically lift, I think the actual implementation change would be pretty significant. I am basically _a priori_ worried about this rabbit hole because it is seemingly one that a group of experts spending person-years of effort working on a datetime library designed to avoid footguns decided to punt on.

# Plausible ideas for the future

I think I generally consider the approach of making units above "seconds" to be of variable length to be implausible. So that leaves us with a few choices:

* Stick with the status quo. We don't deal with leap seconds outside of clamping them when parsing. We don't satisfy any of the three properties I outlined at the beginning of this issue.
* Figure out some other way of making my `TimeScale` approach work. For example, maybe we accept the perf hit of time scale conversions during a time zone lookup. Or perhaps wire it up somehow such that the common case (using the Unix time scale) doesn't do any additional work. Still though, that wasn't the only issue. There are other ergonomic issues I found with this, but maybe we shouldn't treat them as deal-breakers. As outlined above though, this would still not deal with datetime arithmetic in an intuitive way because units below "day" would still be fixed in length. There are also some open questions (for me) for how to tie this into formatting/parsing in a nice way. For example, I believe it would require parsing directly into an `Timestamp<Tai>` or a `Zoned<Tai>` in order to support them, which wouldn't be the default. (Because civil datetimes are, by design, incapable of representing leap seconds.) This seems very non-ideal, because you basically have to encode the knowledge of "supports leap seconds" into the code itself by virtue of using `Timestamp<Tai>` instead of `Timestamp`, _and_ because when you do this, you'll automatically lose intuitive arithmetic semantics whenever you cross a leap second boundary.
* Provide a TAI time zone. This is a [common work-around done for other datetime libraries](https://github.com/ryzokuken/temporal-tai), but is [maybe also a bit odd](https://github.com/tc39/proposal-temporal/issues/900). This seems pretty straight-forward to do by providing TZif data where each transition corresponds to a leap second. Each transition would just add 1 second (since all leap seconds up as of 2024-04-06 are positive). This would provide a way to satisfy property 3 since computing the difference in time between two `Zoned` values already accounts for time zones. But I don't think it would do much more than that. Units of time below "days" would still be fixed, and so dealing with datetime arithmetic even in a hypothetical TAI time zone would still be unintuitive. This would have other downsides, for example, it wouldn't really be able to extend parsing/formatting such that a leap second could be round-tripped. So this would not help to satisfy either property 1 or 2. Moreover, currently, the API of `TimeZone` is defined in terms of `Timestamp` and `civil::DateTime`, neither of which can represent leap seconds.
* Lift the constraint that civil datetimes can't represent leap seconds. This might perhaps open other avenues for exploration, but I think this is a difficult constraint to lift. Namely, civil time is supposed to represent what we see on clocks and is divorced from notions of DST or Earth rotational speed. (To an extent, I suppose.) That is, civil datetimes in my mind correspond to a "purist" interpretation of the Gregorian calendar. Note though that [icu4x seems to take this approach](https://docs.rs/icu_calendar/latest/icu_calendar/types/struct.IsoSecond.html). So it might warrant further exploration.

I overall, at this point, feel like Temporal probably made the best compromise here. We clamp incoming data to something sensible and otherwise ignore leap seconds. I also think that a lot of the trouble and bugs that come from leap seconds aren't necessarily because of datetime libraries not handling them, but rather, because programmers have assumed that their system time is monotonically increasing. Jiff doesn't provide any monotonic guarantees because it's a library for dealing with human time. Monotonic time is something entirely different, and thus, I'm not sure the benefits of handling leap seconds beyond what Temporal does is worth it.

# More on Temporal

I think it's worth capturing some links to more Temporal materials about leap seconds. My intent here was to capture pretty much anything relevant about leap seconds from the Temporal repository/docs/meetings:

* https://github.com/tc39/proposal-temporal/issues/54: is I think the first issue about this where leap second support is rejected entirely. I think my summary here is "because we are tied to POSIX time and POSIX time doesn't support leap seconds." I found this somewhat unsatisfying personally. Indeed, while POSIX specifies that a day is 86,400 seconds, Temporal does not adhere to that constraint. So Temporal has the ability to diverge from POSIX, but why not here?
* https://github.com/tc39/proposal-temporal/issues/339: discusses `constrain` vs `reject` behavior during parsing, but doesn't really discuss leap second support as a whole.
* https://github.com/tc39/proposal-temporal/blob/985b8269306e4ef07ea4e262cf8818493502c6c2/meetings/agenda-minutes-2020-02-13.md?plain=1#L15: Meeting notes that again point out that POSIX says `23:59:60` is the same as `23:59:59`.
* ["We smear in a similar manner to Google, but over 2000 seconds. My personal opinion is that leap seconds have always been, and will probably always be constrained to very niche applications in finance."](https://github.com/tc39/proposal-temporal/issues/1700#issuecomment-1419904659)
* [Should Temporal ship a TAI timezone?](https://github.com/tc39/proposal-temporal/issues/900)
* [RFC 5545 support](https://github.com/tc39/proposal-temporal/issues/759). Notably, Temporal's behavior with respect to leap seconds seems to precisely match what RFC 5545 says.
* ["We decided to ignore leap seconds since they are arbitrarily decided upon and unpredictable."](https://github.com/tc39/proposal-temporal/pull/136/files#diff-87d4c76540f1e15b75fc2c355c38e685004525e11095bc55527e58540d86e343R40) I also find this unconvincing because time zones are also arbitrarily decided upon and unpredictable, yet Temporal supports them.

I've looked through all of these (and looked for additional materials), and I feel like all of the reasons given _as stated_ are not particularly compelling. Instead, I think the real challenge to leap seconds is the following (a distillation of the above exploration):

* In _almost all_ contexts, the source of time (i.e., POSIX time) drops leap seconds before you even have a chance. This in turn makes _representing_ leap seconds in a way that is consistent with the source and without any additional costs impossible.
* Existing standards, like RFC 5545, have been built up around the notion that leap seconds should just be ignored.
* Fitting leap seconds into the design of a higher level datetime library that does the "intuitive" thing is extremely difficult because it seemingly requires both hours and minutes to have a variable length, and whose length is determined relative to a particular datetime. (I did not see this consideration mention in any of the Temporal materials. Perhaps it is implicitly understood by them.)

I find the following reasons to be invalid or entirely unconvincing:

* Leap seconds are not predictable and can change on short notice. As far as I know, this is also true of time zones. Yet, we don't use that as a reason to not support time zones. We've already "solved" the problem of mutable time zone data: environments ship copies of the IANA Time Zone database that most all datetime libraries know how to read. (Although interestingly, the datetime crates in the Rust ecosystem have very spotty support for this. The closest is [`tzfile`](https://crates.io/crates/tzfile) for `chrono` specifically, but its support is incomplete.)
* TAI can be supported as "just another time zone," so leap seconds don't need to be supported in the library. I overall think that this technique is unproven. I've seen what I would consider a proof of concept for this idea, but it's unclear to me if anyone is actually using this technique in production to solve a real problem. Moreover, this approach only addresses property (3) from (way) above, but nothing else as far as I know.

There also seems to be some confusion about what "leap second support" even means. Some people just care about round-tripping. Others just care about computing the accurate difference between two points in time. And yet others care about making the higher level user experience around leap seconds make intuitive sense.

**Comment by mzabaluev** [comment]
> Leap seconds are not predictable and can change on short notice. As far as I know, this is also true of time zones. Yet, we don't use that as a reason to not support time zones. We've already "solved" the problem of mutable time zone data: environments ship copies of the IANA Time Zone database that most all datetime libraries know how to read.

I'd rather say we _have to_ deal with mutable time zone data, because otherwise it's impossible to support time zones with any pretense on correctness. And the issues associated with keeping these data up to date are only resolved in the sunny day case: if your environment does not provide tzdata, or if you fall behind on updates (or lack the capability to receive updates at all), your time arithmetic is liable to go wrong.

Bringing similar issues into computations on seconds is to open another can of worms that I expect very few customers of a time library would want to deal with. In the industry, the predominant thinking has been going in the other direction: push back on any explicit accounting of leap seconds and use smeared NTP time instead.

I've put out a more elaborate write-up on the topic in a [users.rust-lang.org post](https://users.rust-lang.org/t/the-state-of-time-in-rust-leaps-and-bounds/107620); @BurntSushi has read it and commented, but I'm linking here for anyone who might find it useful.

**Comment by BurntSushi** [comment]
> I'd rather say we have to deal with mutable time zone data, because otherwise it's impossible to support time zones with any pretense on correctness.

I think you could say the same about leap seconds though. It just comes down to what kind of precision you attach to notions of "correctness."

> Bringing similar issues into computations on seconds is to open another can of worms that I expect very few customers of a time library would want to deal with. In the industry, the predominant thinking has been going in the other direction: push back on any explicit accounting of leap seconds and use smeared NTP time instead.

I do generally agree, and I think it's ultimately why I ripped leap second support out of Jiff before I released it. It's not a binary thing though. Like, I'm still actually open to some kind of leap second support in Jiff, but it needs to be non-invasive precisely because, like you, I expect very few people to actually want to deal with them.

**Comment by jhpratt** [comment]
>
```
    // For `time`, leap seconds are rejected when parsing
    // via the ISO 8601 format, but are supported via the
    // RFC 2822 format. We want to be maximally charitable
    // here, so pick a format that supports leap seconds.
```

Nit: `time` does support parsing leap seconds. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a585f6e3961acce0d07b1c7e3c029fc

> Conversely, I do not believe the internal design of either chrono or time allows it to satisfy property 3 easily.

In its current state, this is correct (at least for `time`). I do not intend to _ever_ support it with `PrimitiveDateTime`, but `OffsetDateTime` and a future `ZonedDateTime` may.

**Comment by 8573** [comment]
I'm not here to advocate for or against leap second handling, and you clearly have thought more about date/time API design than I have, but I have a minor question or comment.

By "minor" I mean I don't think it will be important in significantly helping you improve this crate, so feel free to ignore it.

---

<details>
<summary>Extended text on which I am commenting</summary>

> However, if our instants accounted for leap seconds, then `2017-03-01T00:00:00Z - 2016-03-01T00:00:00Z` would give you a result of `1 year 1 second`. I frankly find that to be a bit weird, and is analogous to `2024-11-04T00:00:00-05[America/New_York] - 2024-11-03T00:00:00-04[America/New_York]` yielding `1 day 1 hour` instead of the more intuitive and typically desired `1 day`. Indeed, [this is what Temporal does](https://tc39.es/proposal-temporal/docs/zoneddatetime.html#subtract).
> 
> It's worth lingering on this point for a bit, because the way Temporal works here is very intuitive. Namely, when adding units _less_ than a day (hours, minutes, seconds, milliseconds, microseconds, nanoseconds) even when they add up to move than a day, then Temporal will incorporate DST transitions.
>
> [...]
>
> In order to make this work at the user experience level, Temporal had to make any units above "hours" variable length. This is because all time transitions are expressed in units of hours (with a possible fractional component). So once you move above hours, it makes sense to "ignore" the time transitions. For example, if you have an appointment on `2024-03-09T14:00:00-05[America/New_York]` and want to re-schedule it to one day later, then the expected appointment time would be `2024-03-10T14:00:00-04[America/New_York]`. If you implemented that re-scheduling by simply adding 24 hours, then you'd get the wrong result of `2024-03-10T15:00:00-04[America/New_York]`.
> 
> So if we wanted to deal with leap seconds in analogous way, this appears to lead us to the conclusion that any units above _seconds_ need to be of variable length. That means, for example, that adding 1 minute to `2016-12-31T23:59:59Z` should give you `2017-01-01T00:00:59Z`, while adding 60 seconds should give you `2017-01-01T00:00:58Z`. Wild, right?
> 
> From what I can tell, the division between which units are variable (days and higher) and which are not (hours and lower) in Temporal is an architectural constraint the pervades not just its API but also its implementation. For example, if you can assume that all units of hours or below are of a fixed length, [...]
</details>

---

> However, if our instants accounted for leap seconds, then `2017-03-01T00:00:00Z - 2016-03-01T00:00:00Z` would give you a result of `1 year 1 second`. I frankly find that to be a bit weird, and is analogous to `2024-11-04T00:00:00-05[America/New_York] - 2024-11-03T00:00:00-04[America/New_York]` yielding `1 day 1 hour` instead of the more intuitive and typically desired `1 day`.

It seemed to me that this problem, this conflict between two potentially desirable results to which the subtraction should evaluate, stemmed from not distinguishing between calendar time and (making up a term) what one might call physical time either at the type level or with different methods (e.g. `a - b` vs `a.sub_physical(b)`); and this also seems to apply to the tension in the following paragraphs between interpreting units of time span as variable-length vs fixed-length.

I now see that not distinguishing between calendar time and physical time was [a decision explicitly made](https://docs.rs/jiff/latest/jiff/_documentation/design/index.html#why-is-there-one-duration-type-instead-of-two). Though the notes on this crate's design, including trade-offs, are impressively extensive, I don't see that they connect the issues of (1) time differences accounting (or not) for leap seconds and (2) having (or not) separate types for what you term calendar time vs absolute time. Would you agree or disagree that the tension in whether those time differences should be "1 year" or "1 year 1 second" and "1 day 1 hour" or "1 day" is one of the consequences of not distinguishing between calendar and absolute time? (Maybe this was documented as a trade-off and I missed it.)

**Comment by BurntSushi** [comment]
> > 
> 
> ```
>     // For `time`, leap seconds are rejected when parsing
>     // via the ISO 8601 format, but are supported via the
>     // RFC 2822 format. We want to be maximally charitable
>     // here, so pick a format that supports leap seconds.
> ```
> 
> Nit: `time` does support parsing leap seconds in ISO 8601. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a585f6e3961acce0d07b1c7e3c029fc

I think what I meant here is that `time` would fail the test I set out to try. Specifically because the leap second wouldn't be distinct from all other times: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e76fc3c9dbcfd0541cb9ad5ea9f38a80

**Comment by BurntSushi** [comment]
> I now see that not distinguishing between calendar time and physical time was [a decision explicitly made](https://docs.rs/jiff/latest/jiff/_documentation/design/index.html#why-is-there-one-duration-type-instead-of-two). Though the notes on this crate's design, including trade-offs, are impressively extensive, I don't see that they connect the issues of (1) time differences accounting (or not) for leap seconds and (2) having (or not) separate types for what you term calendar time vs absolute time. Would you agree or disagree that the tension in whether those time differences should be "1 year" or "1 year 1 second" and "1 day 1 hour" or "1 day" is one of the consequences of not distinguishing between calendar and absolute time? (Maybe this was documented as a trade-off and I missed it.)

I don't think I got this far in my thinking about leap seconds because this particular caveat of interaction between spans and leap seconds wasn't ultimately the thing that drove me to rip out support. I _think_ I would have lived with it because leap second support still would have been opt in. (That is, the default time scale of `Timestamp` would have been Unix, but you could have opted into TAI.)

In terms of calendar/clock units (or more precisely, non-uniform and uniform units), I think what leap seconds do is put pressure on whether to treat _all_ units above seconds as non-uniform or not. Right now, Jiff always treats hours or smaller as uniform. They always refer to the same absolute duration of time. But if leap seconds were supported, _should_ they still be uniform? I think that would be the key question to address. But either way, I don't think the decision to combine non-uniform and uniform units into one type is that impacted by this. Because any way you slice it, you want calendar units somewhere in your crate API. And you'd still have to answer how leap seconds impact the units on calendar-only spans.

**Comment by ChristopherRabotin** [comment]
Would you be interested in a crate feature called, e.g. "leap-seconds", which would convert a datetime to hifitime's representation to compute the number of leap seconds accumulated for a given time?

I've put in considerable amount of work in ensuring the hifitime is correct, even though properties 1+2 are glitchy in version 3 (and explicitly for +/- one second around the announced leap seconds -- https://github.com/nyx-space/hifitime/issues/255). Yet, the library is considered correct enough for fundamental timing research via GPS spacecraft by the Centre National de Recherches Spatiales (CNRS), and for spacecraft timing calculations by Amazon Ground Systems. I consider this a badge of honor, and I think that hifitime is one of the most precise and thorough time management libraries out there.

In other words, interop between Jiff and Hifitime would be beneficial for the few users of Jiff who need both the time zone management of Jiff and the time scales support of hifitime.

**Comment by BurntSushi** [comment]
I think adding a dependency on `hifitime` from Jiff is definitely not the right way to go. And the reverse seems similarly inappropriate as well. I think a separate crate is the way to go. In #50, I discuss how to approach ecosystem integration more generally, with [this comment in particular](https://github.com/BurntSushi/jiff/issues/50#issuecomment-2253457694) being appealing to me. That doesn't mean jiff<->hifitime integration has to follow the exact same pattern, but "separate crate" is the right idea to follow I think.

Leap second support isn't a priority for me, so if you wanted to pursue publishing a crate that integrates Jiff and hifitime, that might be nice. It might make sense to take the approach of providing extension traits that add methods to Jiff/hifitime data types that allow converting to-and-from the library types. That way, the library isn't leap second focused, but "Jiff and hifitime interop"-focused, if that makes sense.

**Comment by Plecra** [comment]
I'd like to make some tweaks to the docs to help users understand the hazards around time maths, and which operations (like [`Sub`](https://docs.rs/jiff/latest/jiff/struct.Timestamp.html#impl-Sub-for-Timestamp)) are approximations when done on POSIX time. Would a PR on that be welcome?

**Comment by BurntSushi** [comment]
I think it would be better to put that in `_documentation::design` perhaps? Adding warnings about leap seconds to every arithmetic operation between datetimes that returns a duration doesn't sound like a good idea to me personally. If you want to make a more compelling case for it, please open a new issue that discusses the specific documentation changes you'd like to make.

**Comment by autarch** [comment]
FWIW, I implemented support for leap second in the [Perl DateTime.pm module](https://metacpan.org/dist/DateTime) many years ago. I think it satisfies all 3 of your qualifications (more or less).

> Parsing from a string into a Timestamp like 2016-12-31T23:59:60Z (the latest leap second to have occurred as of 2024-
04-06) would correctly represent this time as distinct from all other times, including 2016-12-31T23:59:59Z.

DateTime itself doesn't do parsing. But its constructor will accept leap seconds:

```perl
my $dt = DateTime->new( year => 2016, month => 12, day => 31, hour => 23, minute => 59, second => 60, time_zone => 'UTC' );
```

> Converting an Timestamp back into a string would correctly preserve the leap second.

```perl
my $dt = DateTime->new( year => 2016, month => 12, day => 31, hour => 23, minute => 59, second => 60, time_zone => 'UTC' );
print "$dt\n";

# prints "2016-12-31T23:59:60"

> Computing the difference between two Timestamps would correctly account for leap seconds. For example, 2017-01-01T00:00:00Z - 2016-12-31T23:00:00 would yield 3601 seconds instead of 3600 seconds.

It does this too:

```perl
my $dt1 = DateTime->new( year => 2016, month => 12, day => 31, hour => 23, time_zone => 'UTC' );
my $dt2 = DateTime->new( year => 2017, time_zone => 'UTC' );

my $dur = $dt2->subtract_datetime_absolute($dt1);

print $dur->seconds, "\n";

# prints "3061"
```

And based on all of this what I would say is ...

Definitely don't do this! This was a _huge_ amount of work that I think only makes this library _worse_ in almost all cases. It's just another thing to think about when dealing with datetimes that the vast, vast majority of applications can ignore.

The few cases that care about this sort of stuff, like scientific applications, should be using something much simpler that just represents time as a count of nanoseconds or some other tiny unit. For a general purpose library that attempts to represent the way humans interact with datetimes, leap seconds are unnecessary.

**Comment by BurntSushi** [comment]
Jiff has been out now for a little over 6 months, and I haven't really heard much of a peep about leap seconds. And when I've pointed folks to this issue, the general reaction seems to be, "yeah makes sense."

I'm still open to adding better support for computing durations that account for leap seconds via a special time zone. (Which, to be clear, is technically possible today via `TimeZone::tzif`.) But I'd like to see concrete use cases for that, and specifically why using a crate like `hifitime` isn't possible or desirable.

**Comment by Kinrany** [comment]
I think I had the same thought as @8573 re: distinguishing between calendar time and physical time when thinking about the issues of "1 year 1 day" vs "1 year".

This is likely naive, but the immediate solution I started considering was having two separate `Span` types in the same way there is `Zoned` and the `civil` types.

**Comment by Kinrany** [comment]
I don't have a use case for leap seconds, to be clear: I just found myself reading the issue 😄

And I don't feel strongly about supporting leap seconds, but I feel like it should be _possible_ to do ergonomically, because it does appear to be the same problem as DST on a technical level.

---

Let me take the example from "Status quo":

> For example, you can add things like "years" or even get "years" as a unit of difference between two datetimes. However, if our instants accounted for leap seconds, then `2017-03-01T00:00:00Z - 2016-03-01T00:00:00Z` would give you a result of 1 year 1 second. I frankly find that to be a bit weird, and is analogous to `2024-11-04T00:00:00-05[America/New_York] - 2024-11-03T00:00:00-04[America/New_York]` yielding 1 day 1 hour instead of the more intuitive and typically desired 1 day.

In my mind there are two time systems that exist in parallel here, civil/naive and physical/accurate, and this should apply to durations as well.

`2017-03-01 - 2016-03-01` represents the difference between two civil dates, should result in `1 year`, a civil duration value.

`2024-11-04T00:00:00-05[America/New_York] - 2024-11-03T00:00:00-04[America/New_York]` represents the difference between two physical moments in time, should result in `N nanoseconds`, a physical duration value.

It makes sense to have a canonical conversion between the two duration types. Civil `1 month - 1 day` can be represented as a physical `2.5056 * 10^15 ns` for the purpose of _estimating_ the length. Physical `2505600 * 10^9 ns` can be represented as `29 days` for the purpose of displaying it for humans. However both conversions are lossy, and using them means opting into approximate calculations.

On the other hand, as long as one stays fully within one of the two worlds, all the math can be perfectly precise. So are the conversions of timestamps between the two worlds. By choosing the order of operations, we can express all correct computations while picking and choosing which parts should use calendar math, which parts should be physical, and how much we care about accuracy vs performance vs conciseness.

Two more examples.

Suppose we wanted to calculate `2017-03-01 - 2016-03-01` as a precise physical duration. Instead of calculating the civil duration and then converting it into physical, we can first convert the civil dates into physical timestamps and then calculate the duration between those.

Suppose we wanted to re-schedule `2024-03-09T14:00:00-05[America/New_York]` one day later. We get to convert the physical timestamp into civil time, add `1 day`, convert back into physical and get `2024-03-10T14:00:00-04[America/New_York]`.

In fact it would make perfect sense to define `physical time + civil duration` in this way. `civil time + physical duration` is less obvious to me.

We end up with a matrix of civil/physical, timestamp/duration, and then an extended matrix of all the various conversions between them.

There is a risk of this getting kinda complicated to figure out, but on the other hand if the user is learning the library and does not care about handling leap seconds and perhaps even DST correctly, I like to think it should be possible for them to ignore the whole civil vs physical distinction entirely.

**Comment by BurntSushi** [comment]
I don't think it makes sense to suggest a direction here without concrete use cases.

Also, Jiff already has two duration types, `Span` and `SignedDuration`, that seem to correspond to the duality you're proposing.

Leap seconds aren't a property of the duration. They are a property of the time scale.

The correspondence with DST is something I mentioned above:

>  That's because Unix time is ambiguous with respect to leap seconds, in exactly the same way the civil datetimes are ambiguous with respect to DST transitions.

Also:

> but I feel like it should be possible to do ergonomically

The challenge here is in the API design of how to represent an instant in time that is leap second aware that doesn't completely complicate the API for approximately _every_ Jiff user that won't give a hoot about leap seconds.

**Comment by axelkar** [comment]
Having learned a little bit about leap seconds for learning astronomy, isn't it right that leap seconds are only important when you want to convert the difference between two wall times[^walltime] to literal SI seconds? I feel like you could tell that in the docs.

- `CLOCK_REALTIME` (Used by Rust's `SystemTime`) is nominally in UTC, and it doesn't count leap seconds.

  This means that if you set a timer, the actual count of SI seconds may be different, but if you want to set a timer for when a wall clock says `2025-07-28 07:00`, use this clock.
- `CLOCK_MONOTONIC` (Used by Rust's `Instant`) is the RTC clock, which should theoretically count every SI second (excluding system suspend[^suspend] and hardware imperfections[^monotonic]) like TAI.
- `CLOCK_TAI` is a Linux-specific clock that is offset from `CLOCK_REALTIME`. Since 2017 the offset has been [37 seconds](https://data.iana.org/time-zones/data/leap-seconds.list). Its correct function is dependent on the time daemon telling the kernel when it inserts leap seconds.

You could also link to [`hifitime`'s leap second documentation](https://docs.rs/hifitime/latest/hifitime/index.html#leap-second-support).

[^walltime]: Ignoring DST and timezones, wall time is strictly linked to the rotation of the Earth ([UT1](https://en.wikipedia.org/wiki/Universal_Time) time standard). It's also known as civil time, and UTC and Unix timestamps are a form of wall time.
[^suspend]: `CLOCK_BOOTTIME` is a Linux-specific clock that advances during system suspend.
[^monotonic]: `CLOCK_MONOTONIC` _can_ be skewed by a time daemon; its rate can be adjusted. See [this Stack Overflow answer](https://stackoverflow.com/questions/14270300/what-is-the-difference-between-clock-monotonic-clock-monotonic-raw/14270415#14270415) about `CLOCK_MONOTONIC_RAW`

**Comment by BurntSushi** [comment]
> isn't it right that leap seconds are only important when you want to convert the difference between two wall times[1](https://github.com/BurntSushi/jiff/issues/7#user-content-fn-walltime-bdd7eb60c9002870e177b6cf5f8b9546) to literal SI seconds?

That's only one of the three cases I outlined in the OP.

I also think "wall time" is confusing terminology. In my mind, "wall time" is short for "wall clock time." And that means, "the time on the clocks that you see in your specific location." i.e., civil time. But "wall time" can also be used how you're using it (and I'm certainly guilty of doing that too). So it ends up being somewhat ambiguous.

I do think there is room for improvement in the docs for discussing leap seconds.

> CLOCK_REALTIME (Used by Rust's SystemTime) is nominally in UTC, and it doesn't count leap seconds.

That's POSIX or Unix time, not UTC. "UTC without leap seconds" is just UT1. Leap seconds are the _entire point_ of UTC. :-)

There's a lot of confusion in terminology here, and in a lot of cases, it just doesn't matter. But in this particular context, it very much does.

> CLOCK_MONOTONIC (Used by Rust's Instant) is the RTC clock, which should theoretically count every SI second (excluding system suspend[2](https://github.com/BurntSushi/jiff/issues/7#user-content-fn-suspend-bdd7eb60c9002870e177b6cf5f8b9546) and hardware imperfections[3](https://github.com/BurntSushi/jiff/issues/7#user-content-fn-monotonic-bdd7eb60c9002870e177b6cf5f8b9546)) like TAI.

This is not true. `Instant` only guarantees monotonically increasing values. Each unit of increase isn't guaranteed to be equivalent to other units of increase, nor to any specific definition of SI second.

> You could also link to [hifitime's leap second documentation](https://docs.rs/hifitime/latest/hifitime/index.html#leap-second-support).

`hifitime` is discussed quite a bit in this issue...

**Comment by axelkar** [comment]
> > isn't it right that leap seconds are only important when you want to convert the difference between two wall times[1](https://github.com/BurntSushi/jiff/issues/7#user-content-fn-walltime-bdd7eb60c9002870e177b6cf5f8b9546) to literal SI seconds?
> 
> That's only one of the three cases I outlined in the OP.
> 
> I also think "wall time" is confusing terminology. In my mind, "wall time" is short for "wall clock time." And that means, "the time on the clocks that you see in your specific location." i.e., civil time. But "wall time" can also be used how you're using it (and I'm certainly guilty of doing that too). So it ends up being somewhat ambiguous.

This is the definition I've been using: https://www.w3.org/TR/timezone/#dfn-wall-time

> I do think there is room for improvement in the docs for discussing leap seconds.
> 
> > CLOCK_REALTIME (Used by Rust's SystemTime) is nominally in UTC, and it doesn't count leap seconds.
> 
> That's POSIX or Unix time, not UTC. "UTC without leap seconds" is just UT1. Leap seconds are the _entire point_ of UTC. :-)
> 
> There's a lot of confusion in terminology here, and in a lot of cases, it just doesn't matter. But in this particular context, it very much does.

Thanks for the correction! So UTC counts leap seconds when it jumps back from 23:59:60.999 to 00:00:00.000.

"counting" really is a confusing, ambiguous term. UTC literally counts leap seconds (but jumps back to systems assuming all UTC days are 86400 seconds long), Unix time counts leap seconds (AFAIK with the current behavior) by repeating a second (see contradiction in [`clock_gettime(2) § CLOCK_TAI`](https://man.archlinux.org/man/clock_gettime.2#CLOCK_TAI)), TAI counts seconds that UTC would mark as leap seconds.

Edit: It's not really a contradiction but a reader may imply that the other clocks don't "count" leap seconds.

Your version is simple and correct:

> That's because Unix time is ambiguous with respect to leap seconds

> > CLOCK_MONOTONIC (Used by Rust's Instant) is the RTC clock, which should theoretically count every SI second (excluding system suspend[2](https://github.com/BurntSushi/jiff/issues/7#user-content-fn-suspend-bdd7eb60c9002870e177b6cf5f8b9546) and hardware imperfections[3](https://github.com/BurntSushi/jiff/issues/7#user-content-fn-monotonic-bdd7eb60c9002870e177b6cf5f8b9546)) like TAI.
> 
> This is not true. `Instant` only guarantees monotonically increasing values. Each unit of increase isn't guaranteed to be equivalent to other units of increase, nor to any specific definition of SI second.

Ah, that's right. The standards don't guarantee anything to that extent. It's dependent on time daemons setting the rate from NTP, which isn't a guarantee from the kernel.

> > You could also link to [hifitime's leap second documentation](https://docs.rs/hifitime/latest/hifitime/index.html#leap-second-support).
> 
> `hifitime` is discussed quite a bit in this issue...

I was thinking about the docs section linked below but `hifitime` is mentioned above it:
https://docs.rs/jiff/latest/jiff/_documentation/design/index.html#why-doesnt-jiff-support-leap-seconds

**Comment by BurntSushi** [comment]
> This is the definition I've been using: https://www.w3.org/TR/timezone/#dfn-wall-time

Yeah that's the one I referenced. But the surrounding context of your words suggested you were using absolute or physical time. Like I wouldn't say that Unix timestamps are an example of wall time. Unix timestamps are a measure of physical time. Unix timestamps aren't what you see on analog clocks.

> "counting" really is a confusing, ambiguous term. UTC literally counts leap seconds (but jumps back to systems assuming all UTC days are 86400 seconds long), Unix time counts leap seconds (AFAIK with the current behavior) by repeating a second (see contradiction in [clock_gettime(2) § CLOCK_TAI](https://man.archlinux.org/man/clock_gettime.2#CLOCK_TAI)), TAI and UT1 count leap seconds by ignoring them.

Repeating a second occurs when a second is _added_ to a day. That's a positive leap second. All leap seconds up to this point have been positive. But there's no reason why a negative leap second couldn't also be added (unless leap seconds are abolished first, which is possible, since it has been "decided" to abandon leap seconds on or before 2035). In the case of a negative leap second, you go from `23:59:58.999` to `00:00:00`.

Saying that TAI and UT1 count by ignoring leap seconds is a correct thing to say. But it's not how I'd phrase it. A leap second is the defining concept of UTC and is how it is related to UT1 and TAI. So I see it more as "UTC is TAI plus leap seconds" as opposed to "TAI is UTC without leap seconds." Leap seconds are a funny "human" concept that just doesn't have a place in something like a physical measurement like TAI. (At a surface level I suppose. At a deeper level, even things like TAI are just funny human concepts tied to our perception of reality.)

The way I personally think about it is that UTC is a _model_ of time that tries to balance physical measurement (rotation of the earth, 1 SI second) with human concerns ("noon" is roughly when the sun is at its apex).

> I was thinking about the docs section linked below but hifitime is mentioned above it:
https://docs.rs/jiff/latest/jiff/_documentation/design/index.html#why-doesnt-jiff-support-leap-seconds

Oh I see. I misunderstood. Yeah I think that would be fine. I gave a very short answer in the docs and just pointed to this issue. But it would make sense to collate this issue into one short digestible summary. It's just that doing so is difficult, so I punted. :-)

**Comment by ChristopherRabotin** [comment]
Point of clarification : UT1 isn't monotonic at a constant rate, whereas TAI is, and UTC is not. UTC will stop counting when a leap second occurs for it to catch up. The point of UTC is to roughly define noon as when the sun is at the zenith twice a year at Greenwich. UT1 is a measure of time with respect to an inertial (ie non accelerating) frame defined by distant stars that we consider to be non accelerating. Leap seconds therefore round the error between UT1 and UTC to ensure that both time scales are close enough. The duration of one day in UT1 is not 24 hours (or 23h 56 min 4s, as a sidereal day): UT1 isn't predictable.

**Comment by axelkar** [comment]
> absolute or physical time [...] wall time

Which is which? I believe wall time to be the same as civil time and what a watch shows. Do you mean TAI by physical time?

> > convert the difference between two wall times to literal SI seconds

This means converting a span of time in UTC (wall time/civil time) to a number of seconds of TAI. That is — adding leap seconds.

> > wall clock says `2025-07-28 07:00`

Same thing here. You should sleep for a number of UTC seconds, not TAI seconds.

> [you're only talking about positive leap seconds]

Yes! Important correction! I couldn't apply that definition of "counting" to negative leap seconds.

> The way I personally think about it is that UTC is a model of time that tries to balance physical measurement (rotation of the earth, 1 SI second) with human concerns ("noon" is roughly when the sun is at its apex).

Aren't _rotation of the Earth_ and _noon_ the same thing here? Humans want solar time, which is dependent on the rotation of the Earth.

> Point of clarification : UT1 isn't monotonic at a constant rate, whereas TAI is, and UTC is not. UTC will stop counting when a leap second occurs for it to catch up.

From what I've heard, UTC days can have 86399, 86400 or 86401 seconds, and I've thought that on positive leap seconds UT C counts all the way to 86400.999... seconds before beginning the next day from 0.

> The duration of one day in UT1 is not 24 hours (or 23h 56 min 4s, as a sidereal day): UT1 isn't predictable.

Yep, and that's why we can't predict leap seconds.

Side note: Do sidereal days not change in duration when the Earth's axial rotation slows down?

I think Jiff should also include a warning for the uninformed that astronomy isn't necessarily done with UTC, especially if there are functions related to Julian days.

As a counterpoint to that, calculating sunrise equations isn't all that accurate and you're outputting with a precision of minutes: Going back from the `J2000` reference point by `30 * 365,25` days isn't so bad for converting to/from Unix time. With my napkin math leap seconds are somewhere below a part per million, if not parts per billion.

**Comment by BurntSushi** [comment]
UTC is not civil time. It also isn't a time zone either, although it can be "used" as such in many contexts. UTC is a time scale that represents physical time. Unix time is also a time scale that represents physical time. Looking at my clock right now, it says 16:10 on July 27, 2025. That is civil/wall/local/naive time.

**Comment by ChristopherRabotin** [comment]
> Humans want solar time, which is dependent on the rotation of the Earth.

That's debatable ;-) Humans want predictable time, with a slight preference for not being obviously completely off from solar time (cf. daylight savings, only nine timezones in Russia, the same time from Poland to Spain, etc.).

> Do sidereal days not change in duration when the Earth's axial rotation slows down?

Yes, but it is negligible on a human time scale compared to the rotation of the Earth on itself (you can use [ANISE](https://nyxspace.com/anise/) to query the daily NASA publication of the high precision Earth rotation parameters). Sidereal time is 24 hours minus 1/365.25 days (which is how much the Earth progresses in its orbit around the Sun in one day):

```
>>> from hifitime import *
>>> Unit.Hour*24 - Unit.Day * (1/365.25)
23 h 56 min 3 s 449 ms 691 μs 992 ns
```

> Going back from the J2000 reference point by 30 * 365,25 days isn't so bad for converting to/from Unix time

For astronomy or spaceflight dynamics, I strongly recommend that you just use hifitime. It's NASA Technology Readiness Level 9, meaning it's been used for a successful spacecraft mission; in the case of hifitime, it was used on the 100-million dollar Blue Ghost lunar lander, the first successful lunar lander by an American team in two generations since 1972. Hifitime is also used in astronomy and for fundamental GNSS timing research (GPS, Galileo, etc.). As a professional in spaceflight dynamics and the in-house time system expect at a important new space company, I can't insist enough that you should just use hifitime for time scale conversions instead of rewriting your own. Every year I'm brought on to consult on some project where people tried to redevelop the leap second or time system conversions thinking it just wasn't that hard, break their teeth, and waste hundreds of engineering hours trying to fix it. Just ten days ago I fixed a time conversion bug for a customer simply by replacing all of the custom conversions to use hifitime. Would you rewrite a linear algebra library for production system? If the answer is _no, I would just use nalgebra/numpy_, then don't rewrite time system conversions and leap second handling, just use hifitime. 

I'll end my rant with my favorite quote from SOFA:

>  Leap seconds pose tricky problems for software writers, and consequently there are concerns that these events put safety-critical systems at risk. The correct solution is for designers to base such systems on TAI or some other glitch-free time scale, not UTC, but this option is often overlooked until it is too late.

-- "SOFA Time Scales and Calendar Tools", Document version 1.61, section 3.5.1

https://nyxspace.com/hifitime/#time-scales

**Comment by axelkar** [comment]
> > This means converting a span of time in UTC (wall time/civil time) to a number of seconds of TAI.
>
> UTC is not civil time. It also isn't a time zone either, although it can be "used" as such in many contexts. [...] Looking at my clock right now, it says 16:10 on July 27, 2025. That is civil/wall/local/naive time.

I meant a span of time in Unix time (calendar abstractions are definitely wall time, and are trivially implemented on top of Unix time, but anyways I mean seconds of Unix time), not UTC. You're right.

I think you also mean that floating/naive/local time isn't identical to civil/wall time, but that your clock has both traits.

> UTC is a time scale that represents physical time. Unix time is also a time scale that represents physical time.

Physical time as in a count of SI seconds? For UTC - the time scale, not the time zone - it's true that it counts all seconds. ... I just had an epiphany. UTC _uniquely identifies_ (replacement to ambiguous "counts") all seconds, whereas Unix time doesn't. That translates to UTC not having ghost times, and it's true, but how do I link that back to leap seconds? I'm getting side tracked by terminology. Anyways, I mean that Unix time can't represent all points in time unlike UTC (which can't do it into the future).

As such, I think the second sentence is incorrect.

> you can use [ANISE](https://nyxspace.com/anise/) to query the daily NASA publication of the high precision Earth rotation parameters

That's amazing!

> I can't insist enough that you should just use hifitime for time scale conversions instead of rewriting your own.

Of course! I've just been reading and trying to understand some undocumented libraries. I actually read the page you linked just a week ago.

**Comment by BurntSushi** [comment]
> > > This means converting a span of time in UTC (wall time/civil time) to a number of seconds of TAI.
> > 
> > 
> > UTC is not civil time. It also isn't a time zone either, although it can be "used" as such in many contexts. [...] Looking at my clock right now, it says 16:10 on July 27, 2025. That is civil/wall/local/naive time.
> 
> I meant a span of time in Unix time (calendar abstractions are definitely wall time, and are trivially implemented on top of Unix time, but anyways I mean seconds of Unix time), not UTC. You're right.
> 
> I think you also mean that floating/naive/local time isn't identical to civil/wall time, but that your clock has both traits.
> 
> > UTC is a time scale that represents physical time. Unix time is also a time scale that represents physical time.
> 
> Physical time as in a count of SI seconds? For UTC - the time scale, not the time zone - it's true that it counts all seconds. ... I just had an epiphany. UTC _uniquely identifies_ (replacement to ambiguous "counts") all seconds, whereas Unix time doesn't. That translates to UTC not having ghost times, and it's true, but how do I link that back to leap seconds? I'm getting side tracked by terminology. Anyways, I mean that Unix time can't represent all points in time unlike UTC (which can't do it into the future).
> 
> As such, I think the second sentence is incorrect.

Physical or "absolute" time is just referring to a concept that describes an _instant in time that is the same for everyone everywhere_. So, UT1, UTC, TAI and Unix time are all notions of absolute time.

See https://github.com/golang/go/issues/19700#issuecomment-559250634 and https://errorprone.info/docs/time for more elaboration.

This is quite distinct from civil/naive/local/wall time, which cannot be mapped to _any_ universal instant in time without some additional information (e.g., a time zone).

To be honest, I don't think there is a widely accepted precise definition of "physical" or "absolute" time. I think it's more a colloquial or casual phrasing meant to draw a distinction between a timestamp and the time that humans see on their clocks.
